### PR TITLE
add /search_players api route

### PR DESCRIPTION
**Pull request by 0i8**
[works well on my server](https://api.halcyon.moe/search?src=j), thought it would be beneficial enough to bring to base bpy

written together with [syswhite](https://github.com/SysWhiteDev)

### CODE REVIEW

File: app/api/domains/api.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ async def api_search(
     # assign placeholders (bindparams()) to these queries,
     # and this will let us bind those safely and efficiently.
     results_query = text(
```
Comment (minisbett) [inline app/api/domains/api.py:123]: You also don't need this whole query at all, you can just all len() on the resulting list of users.

**Review by minisbett — CHANGES_REQUESTED** [review]

File: app/api/domains/api.py [hunk 2]
```diff
@@ -107,6 +109,40 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
         "diff": m.diff,
     }
 
+@router.get("/search")
+async def api_search(
+    search: Optional[str] = Query(None, alias='src'),
```
[the remark this answers is not in the document]
Comment (7ez) [reply app/api/domains/api.py]: "query" or "q" would be more preferable imo

File: app/api/domains/api.py [hunk 3]
```diff
@@ -107,6 +109,40 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+    search_clause = f"%{search}%" if search else None
+    results_query = results_query.bindparams(search_clause=search_clause)
+    rows_query = rows_query.bindparams(search_clause=search_clause)
+
+    # execute the queries using the database connection
+    results = await db_conn.fetch_val(results_query)
+    rows = await db_conn.fetch_all(rows_query)
+
+    # return the response
+    response = ORJSONResponse(
```
Comment (7ez) [inline app/api/domains/api.py]: why are you assigning the response instead of just retuurning it?

File: app/api/domains/api.py [hunk 4]
```diff
@@ -107,6 +109,40 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+    # assign placeholders (bindparams()) to these queries, 
+    # and this will let us bind those safely and efficiently.
+    results_query = text("SELECT COUNT(id) FROM users WHERE name AND priv >= 3 LIKE :search_clause") # we will only need to lookup users who have Privileges.VERIFIED
+    rows_query = text("SELECT id, name FROM users WHERE name LIKE :search_clause AND priv >= 3 ORDER BY id ASC") # to save on host resources and prevent useless requests
+
+    # using parameterised queries helps prevent against sql injections
+    # and will also allow the database engine to cache our queries,
+    # making these lookups faster in the future
+    search_clause = f"%{search}%" if search else None
+    results_query = results_query.bindparams(search_clause=search_clause)
```
Comment (7ez) [inline app/api/domains/api.py:134]: why not let the database handle parameter binding?

File: app/api/domains/api.py (hunk 3 above)
Comment (0i8) [reply app/api/domains/api.py]: this was already changed in the last commit

Comment (minisbett) [reply app/api/domains/api.py]: Check out the code I sent in general on the Discord, I'd suggest replacing this whole PR with that code

Comment (7ez) [reply app/api/domains/api.py]: my code* :trollface:

Comment (minisbett) [reply app/api/domains/api.py]: https://discord.com/channels/748687781605408908/748687781605408911/1056576143668555907

Comment (minisbett) [reply app/api/domains/api.py]: > my code* :trollface:

Didn't say it's mine

File: app/api/domains/api.py (hunk 4 above)
Comment (0i8) [reply app/api/domains/api.py:134]: using bindparams() lets me bind the values safely, for when i am constructing queries dynamically (such as in this route). this is the safest way to avoid SQL injection, and improves performance on search because the engine will cache our queries.

in contrast, concatenating values directly into the query string can open us up for SQL injection, and makes our queries slower and uncacheable. if you have a different approach that avoids the latter, i would love to hear it

File: app/api/domains/api.py (hunk 3 above)
Comment (0i8) [reply app/api/domains/api.py]: no

File: app/api/domains/api.py (hunk 4 above)
Comment (minisbett) [reply app/api/domains/api.py:134]: Check the code on the server...

File: app/api/domains/api.py (hunk 1 above)
Comment (0i8) [reply app/api/domains/api.py:123]: len() is slower by 50+ms on every query if i used it. 

count() is more efficient because it returns the number of rows in the condition specified without having to retrieve all of the rows after that query is done, and then count those rows.

File: app/api/domains/api.py (hunk 4 above)
Comment (7ez) [reply app/api/domains/api.py:134]: you can just do
```py
await db_conn.fetch_val("SELECT name FROM users WHERE id = :id", {"id":1})
```
for example

File: app/api/domains/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (0i8) [reply app/api/domains/api.py]: ok

File: app/api/domains/api.py (hunk 1 above)
Comment (7ez) [reply app/api/domains/api.py:123]: > len() is slower by 50+ms on every query if i used it. 
> 
> count() is more efficient because it returns the number of rows in the condition specified without having to retrieve all of the rows after that query is done, and then counting those rows.

idk what you're talking about, I ran a timeit test on `len([1, 2, 3] * 100)` and got the result of ```2000000 loops, best of 5: 6.26 usec per loop```

File: app/api/domains/api.py (hunk 4 above)
Comment (0i8) [reply app/api/domains/api.py:134]: i don't agree with this approach, because in most cases with this route, we're executing multiple queries handled by a frontend where `fetch_val()` is slower overall than `bindparams()`

if we're going to be dynamically constructing a query, then `bindparams()` is the way i want to do this.
if we are just going to be executing a singular query with fixed values, then i would agree with `fetch_val()`

Comment (7ez) [reply app/api/domains/api.py:134]: > i don't agree with this approach, because in most cases with this route, we're executing multiple queries handled by a frontend where `fetch_val()` is slower overall than `bindparams()`
> 
> if we're going to be dynamically constructing a query, then `bindparams()` is the way i want to do this.
> if we are just going to be executing a singular query with fixed values, then i would agree with `fetch_val()`
> 
> both could be equally good towards this, i would like some more input before i make any changes though

fetch_val was an example. It works on any fetch or execute method.

Comment (minisbett) [reply app/api/domains/api.py:134]: Then rewrite all of bpy so everything is uniform again

File: app/api/domains/api.py (hunk 1 above)
Comment (0i8) [reply app/api/domains/api.py:123]: the implementation of len() for this wouldn't work on my machine and only returned no values every time, idk what it thinks its doing lol

File: app/api/domains/api.py (hunk 4 above)
Comment (cmyui) [reply app/api/domains/api.py:134]: hey, the databases library were using is already doing this stuff behind the scenes for you - no need to use it directly here

File: app/api/domains/api.py [hunk 5]
```diff
@@ -38,6 +39,7 @@
 # or keep up with changes to https://github.com/JKBGL/gulag-api-docs.
 
 # Unauthorized (no api key required)
+# GET /search: returns a list of matching users, based on a passed string, sorted by ascending ID.
```
Comment (cmyui) [inline app/api/domains/api.py:42]: ideally, this should be something like `GET /players`, but since the whole API is currently wrong, let's keep it consistently wrong 🥲 
```suggestion
# GET /get_players: returns a list of matching users, based on a passed string, sorted by ascending ID.
```

File: app/api/domains/api.py [hunk 6]
```diff
@@ -108,6 +110,53 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
     }
 
 
+@router.get("/search")
```
Comment (cmyui) [inline app/api/domains/api.py:113]:
```suggestion
@router.get("/get_players")
```

File: app/api/domains/api.py [hunk 7]
```diff
@@ -108,6 +110,53 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+async def api_search(
```
Comment (cmyui) [inline app/api/domains/api.py:114]:
```suggestion
async def api_search_players(
```

File: app/api/domains/api.py (hunk 5 above)
Comment (cmyui) [reply app/api/domains/api.py:42]: this puts me thru great pain

File: app/api/domains/api.py [hunk 8]
```diff
@@ -108,6 +110,53 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+
+    # we're using text() here because we are going to
+    # assign placeholders (bindparams()) to these queries,
+    # and this will let us bind those safely and efficiently.
+    results_query = text(
+        "SELECT COUNT(id) FROM users WHERE name LIKE :search_clause AND priv >= 3",
+    )  # we will only need to lookup users who have Privileges.VERIFIED
+    rows_query = text(
+        "SELECT id, name FROM users WHERE name LIKE :search_clause AND priv >= 3 ORDER BY id ASC",
+    )  # to save on host resources and prevent useless requests
```
Comment (cmyui) [inline app/api/domains/api.py:128]: bit more correct
```suggestion
        "SELECT COUNT(id) FROM users WHERE name LIKE :search_clause AND priv & 3 = 3",
    )  # we will only need to lookup users who have Privileges.VERIFIED
    rows_query = text(
        "SELECT id, name FROM users WHERE name LIKE :search_clause AND priv & 3 = 3 ORDER BY id ASC",
    )  # to save on host resources and prevent useless requests
```

**Comment by cmyui** [comment]
aw man, this just reminds me how bad the current API is

File: app/api/domains/api.py (hunk 4 above)
Comment (0i8) [reply app/api/domains/api.py]: sure, we'll use it

File: app/api/domains/api.py [hunk 9]
```diff
@@ -108,6 +109,38 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+async def api_search_players(
+    search: Optional[str] = Query(None, alias="q", min=2, max=32),
+    db_conn: databases.core.Connection = Depends(acquire_db_conn),
+):
+    """Search for users on the server by name."""
+
+    # execute the query using the database connection
+    rows = await db_conn.fetch_all(
+        "SELECT id, name FROM users WHERE name LIKE :search AND priv & 3 = 3 ORDER BY id ASC",
+        {"search": f"%{search}%"},
```
Comment (cmyui) [inline app/api/domains/api.py:122]: not sure this handles the None case correctly at the moment - i think we'd want 
```suggestion
        {"search": f"%{search}%" if search is not None else None},
```

Comment (cmyui) [reply app/api/domains/api.py:122]: if you could test this, would be epic

File: app/api/domains/api.py [hunk 10]
```diff
@@ -108,6 +109,38 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
 
 
+@router.get("/search_players")
+async def api_search_players(
+    search: Optional[str] = Query(None, alias="q", min=2, max=32),
+    db_conn: databases.core.Connection = Depends(acquire_db_conn),
+):
+    """Search for users on the server by name."""
+
+    # execute the query using the database connection
```
Comment (cmyui) [inline app/api/domains/api.py:119]: code is self explanatory here
```suggestion
```

File: app/api/domains/api.py (hunk 9 above)
Comment (cmyui) [reply app/api/domains/api.py:122]: maybe we can just disallow empty query; might be simpler

File: app/api/domains/api.py (hunk 10 above)
Comment (cmyui) [inline app/api/domains/api.py:119]: i used to put comments like this all over my code, but my new mindset is https://www.youtube.com/watch?v=2a_ytyt9sf8&t=626s

Comment (cmyui) [reply app/api/domains/api.py:119]: the code here is actually very self-expressive though, i think it's quite readable without the comments

Comment (0i8) [reply app/api/domains/api.py:119]: while i don't agree with this mindset, i will apply it here

Comment (cmyui) [reply app/api/domains/api.py]: would recommend actually watching into the vid though, he explains it more in-depth - e.g. refactoring your code to read *as if it were a comment*;
image

File: app/api/domains/api.py [hunk 11]
```diff
@@ -108,6 +109,35 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+        "SELECT id, name FROM users WHERE name LIKE :search AND priv & 3 = 3 ORDER BY id ASC",
+        {"search": f"%{search}%" if search is not None else None},
+    )
+
+    if len(rows) == 0:
+        return ORJSONResponse(
+            {
+                "status": "error",
+                "message": "nothing was found matching those parameters!",
+            },
```
Comment (cmyui) [inline app/api/domains/api.py:128]: hmm, seems this is inconsistent with my old style - scrolling down i see some like
```py
return ORJSONResponse(
    {"status": "Invalid gamemode."},
    status_code=status.HTTP_400_BAD_REQUEST,
)
```
i actually prefer what you have lol, it's what i use in my newer projects

quite unfortunate

File: app/api/domains/api.py (hunk 10 above)
Comment (0i8) [reply app/api/domains/api.py]: this makes more sense, i'll watch the video when i get a chance

File: app/api/domains/api.py (hunk 11 above)
Comment (0i8) [reply app/api/domains/api.py:128]: pre-commit changed that, originally it was in line with what you have already

Comment (cmyui) [inline app/api/domains/api.py:128]:
```suggestion
            {"status": "Nothing was found matching those parameters."},
```

Comment (cmyui) [reply app/api/domains/api.py:128]: actually, on second thought, let's return 200 with [] result

File: app/api/domains/api.py [hunk 12]
```diff
@@ -108,6 +109,35 @@ def format_map_basic(m: Beatmap) -> dict[str, object]:
...
+            status_code=status.HTTP_404_NOT_FOUND,
+        )
+    else:
+        return ORJSONResponse(
+            {
+                "status": "success",
+                "results": len(rows),
+                "result": [dict(row) for row in rows],
+            },
+        )
```
Comment (cmyui) [inline app/api/domains/api.py:138]:
```suggestion
    return ORJSONResponse(
        {
            "status": "success",
            "results": len(rows),
            "result": [dict(row) for row in rows],
        },
    )
```

**Review by cmyui — CHANGES_REQUESTED** [review]
need to test null case before i can merge this

File: app/api/domains/api.py (hunk 12 above)
Comment (0i8) [reply app/api/domains/api.py]: why always return 200?

Comment (7ez) [reply app/api/domains/api.py]: no point in doing more serverside work

Comment (cmyui) [reply app/api/domains/api.py]: this is more standard in rest

when fetching an array of resources, an empty array is still a valid response
an example might be, e.g. `GET /users` returning `200 OK` with body `[]`

404 is more for the case where you're looking up a specific resource and it is not found
an example might be, e.g. `GET /users/1` returning `404 NOT FOUND`

**Comment by 0i8** [comment]
> need to test null case before i can merge this

what is the intended behaviour? looks and works fine in all cases it appears, on my end

**Comment by cmyui** [comment]
> > need to test null case before i can merge this
> 
> what is the intended behaviour? looks and works fine in all cases it appears, on my end

just wasn't sure the null case would work - i've got a setup running now, i'll get to this shortly np

**Comment by cmyui** [comment]
returns all players upon null query now, as i'd expect

**Review by cmyui — APPROVED** [review]
+1 👍🏻

**Comment by 0i8** [comment]
img
